### PR TITLE
KAFKA-12577; Remove deprecated `ConfigEntry` constructor for 3.0

### PR DESCRIPTION
**Pull request by dajac**
`ConfigEntry`'s constructor was deprecated in 1.1.0. It seems that we can safely remove it now.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dajac** [comment]
I need to file a Jira and update the upgrade nodes.

File: docs/upgrade.html [hunk 1]
```diff
@@ -28,6 +28,8 @@ <h5><a id="upgrade_300_notable" href="#upgrade_300_notable">Notable changes in 3
     <li>The Streams API removed all deprecated APIs that were deprecated in version 2.5.0 or earlier.
         For a complete list of removed APIs compare the detailed Kafka Streams upgrade notes.</li>
     <li>The deprecated <code>Metric#value()</code> method was removed (<a href="https://issues.apache.org/jira/browse/KAFKA-12573">KAFKA-12573</a>).</li>
+    <li>The deprecated <code>org.apache.kafka.clients.admin.ConfigEntry</code> constructor was removed (<a href="https://issues.apache.org/jira/browse/KAFKA-12577">KAFKA-12577</a>).
```
Comment (ijuma) [inline docs/upgrade.html]: Do we need the fully qualified name? We didn't use it for the `Metric` case.

**Review by ijuma — APPROVED** [review]
LGTM, one nit below.

File: docs/upgrade.html (hunk 1 above)
Comment (dajac) [reply docs/upgrade.html]: We have multiple `ConfigEntry` classes in the code base so I used the fully qualified name to be precise. `Metric` is unique however. Shall we use the fully qualified name for `Metric` as well in order to be consistent?

Comment (ijuma) [reply docs/upgrade.html]: Do we have multiple `ConfigEntry` that are public APIs?  One option would be to link to the file in GitHub using a commit hash from the 2.8 branch. That would be both concise and informative (the deprecation documentation would include the suggested replacement typically).

Comment (ijuma) [reply docs/upgrade.html]: On the other hand, it won't be practical to do the same for cases where there are a lot of deprecated usages. So, maybe we should keep it simple and people should make sure to remove deprecations before the upgrade so they get useful help.

Comment (dajac) [reply docs/upgrade.html]: > Do we have multiple `ConfigEntry` that are public APIs?

Good point. There is only one in the public APIs. Let me remove the fully qualified name.

**Comment by dajac** [comment]
@ijuma Updated the PR.

**Review by ijuma — APPROVED** [review]
LGTM

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by dajac** [comment]
Failed tests are not related to this change. Merging to trunk.

**Comment by sknot-rh** [comment]
I am using this deprecated ctor in my tests. How can I set, for example `source` option now? The other ctor has a package protected access.
```
ConfigEntry(String name, String value, ConfigSource source, boolean isSensitive, boolean isReadOnly,
                List<ConfigSynonym> synonyms, ConfigType type, String documentation) 
```

// edit
I used reflection to invoke protected ctor.

**Comment by chia7712** [comment]
> I am using this deprecated ctor in my tests

just curious. Why you need to create `ConfigEntry` in your tests? Could you share the use case to me? thanks

**Comment by sknot-rh** [comment]
The test scenario is basically "mocking" server response to `describeConfig`. Then I am looking at the `source` property to determine whether the changed configuration property is default (`source==DEFAULT_CONFIG) or custom (`source=UNKNOWN`).

**Comment by chia7712** [comment]
> The test scenario is basically "mocking" server response to describeConfig.

Do you create a mock kafka admin and then mock response for specify request. If so, it seems to me you can mock the `ConfigEntry` according request and then use the mocked `ConfigEntry` to create response.

**Comment by sknot-rh** [comment]
No, it is just a junit test consuming properties in `key=value` format. And I want to mark init configuration as default.

**Comment by chia7712** [comment]
ok. How about using reflection to update private fields of `ConfigEntry`? For example:

```java
        Class<?> clazz = ConfigEntry.class;
        ConfigEntry entry = new ConfigEntry("name", "value");
        Field f1 = entry.getClass().getDeclaredField("source");
        f1.setAccessible(true);
        f1.set(entry, ConfigEntry.ConfigSource.DYNAMIC_BROKER_CONFIG);
```

**Comment by sknot-rh** [comment]
Yes, I used something similar. Thank you for looking :)

**Comment by ijuma** [comment]
I think it makes sense to make the constructor public, but it would require a small kip. We don't gain anything by not making it public if people access it via reflection anyway.

**Comment by sknot-rh** [comment]
Let me know if I can do my first contribution! :laughing:

**Comment by tombentley** [comment]
[KIP-692](https://cwiki.apache.org/confluence/display/KAFKA/KIP-692:+Make+AdminClient+value+object+constructors+public) proposed to do this for all the record-like classes in the admin client. I don't think it got any votes at the time.

**Comment by ijuma** [comment]
@tombentley maybe we should vote for it then :)

**Pull request merged**